<a href="https://colab.research.google.com/github/Toulouse08/IS-170-Machine-Learning/blob/Association-Rules/Top10anime_170.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
from mlxtend.frequent_patterns import apriori, association_rules


In [ ]:
df=pd.read_csv('Top10anime.csv')
df.head()

One Piece Action, Adventure, Comedy, Drama, Fantasy, Shounen
0       Blue Lock                     Action, Drama, Shounen, Sports
1   Solo Leveling                Action, Adventure, Fantasy, Shounen
2  Jujutsu Kaisen  Action, Drama, Horror, School Life, Shounen, S...
3         Berserk  Action, Adventure, Fantasy, Horror, Mature, Se...
4   One-Punch Man  Action, Adventure, Comedy, Drama, Sci-fi, Sein...

In [ ]:
#Since it's missing a header we'll create one, the new column header 'Products' 
df=pd.read_csv('Top10anime.csv', names=['tags'])
df.head() 


tags
One Piece       Action, Adventure, Comedy, Drama, Fantasy, Sho...
Blue Lock                          Action, Drama, Shounen, Sports
Solo Leveling                 Action, Adventure, Fantasy, Shounen
Jujutsu Kaisen  Action, Drama, Horror, School Life, Shounen, S...
Berserk         Action, Adventure, Fantasy, Horror, Mature, Se...

In [ ]:
df.shape

(10, 1)

In [ ]:
#the following code converts a pandas dataframe into a list of lists.
data=list(df["tags"].apply(lambda x:x.split(",")))
data

[['Action', ' Adventure', ' Comedy', ' Drama', ' Fantasy', ' Shounen'],
 ['Action', ' Drama', ' Shounen', ' Sports'],
 ['Action', ' Adventure', ' Fantasy', ' Shounen'],
 ['Action', ' Drama', ' Horror', ' School Life', ' Shounen', ' Supernatural'],
 ['Action',
  ' Adventure',
  ' Fantasy',
  ' Horror',
  ' Mature',
  ' Seinen',
  ' Supernatural',
  ' Tradegy'],
 ['Action',
  ' Adventure',
  ' Comedy',
  ' Drama',
  ' Sci-fi',
  ' Seinen',
  ' Supernatural'],
 ['Action',
  ' Comedy',
  ' Drama',
  ' Horror',
  ' Mature',
  ' Shounen',
  ' Supernatural'],
 ['Action',
  ' Comedy',
  ' Drama',
  ' Fantasy',
  ' School Life',
  ' Sci-fi',
  ' Shounen',
  ' Supernatural'],
 ['Action',
  ' Adventure',
  ' Drama',
  ' Historical Martial Arts',
  ' Mature Seinen'],
 ['Action', ' Adventure', ' Drama', ' Fantasy', ' Shounen', ' Supernatural']]

In [ ]:
#Let's transform the list with one-hot encoding (make sure you understand the one-hot encoding concept first)
#Each column represents an item and each row represent a record or a transaction. 
#Initially, if that particular item occurs then it's labeled TRUE, if not then FALSE.
#After that, TRUE is converted to 1 and FALSE to 0 because computer loves numeric values.

from mlxtend.preprocessing import TransactionEncoder
a=TransactionEncoder()
a_data=a.fit(data).transform(data)
df=pd.DataFrame(a_data, columns=a.columns_)

#Turn False & True to 0 & 1 for easy readability
df=df.replace(False,0)
df=df.replace(True,1)

df
                

Adventure   Comedy   Drama   Fantasy   Historical Martial Arts   Horror  \
0           1        1       1         1                         0        0   
1           0        0       1         0                         0        0   
2           1        0       0         1                         0        0   
3           0        0       1         0                         0        1   
4           1        0       0         1                         0        1   
5           1        1       1         0                         0        0   
6           0        1       1         0                         0        1   
7           0        1       1         1                         0        0   
8           1        0       1         0                         1        0   
9           1        0       1         1                         0        0   

    Mature   Mature Seinen   School Life   Sci-fi   Seinen   Shounen   Sports  \
0        0               0             0        0        0         1        0   
1        0               0             0        0        0         1        1   
2        0               0             0        0        0         1        0   
3        0               0             1        0        0         1        0   
4        1               0             0        0        1         0        0   
5        0               0             0        1        1         0        0   
6        1               0             0        0        0         1        0   
7        0               0             1        1        0         1        0   
8        0               1             0        0        0         0        0   
9        0               0             0        0        0         1        0   

    Supernatural   Tradegy  Action  
0              0         0       1  
1              0         0       1  
2              0         0       1  
3              1         0       1  
4              1         1       1  
5              1         0       1  
6              1         0       1  
7              1         0       1  
8              0         0       1  
9              1         0       1

In [ ]:
#We'll try 0.2 for the SUPPORT threshold value (but you can set it to any value you want).
# The parameter (use_colnames=True) is used to convert the integer back to its respective name.

df=apriori(df,min_support=0.2,use_colnames=True)
df

support                                           itemsets
0        0.6                                       ( Adventure)
1        0.4                                          ( Comedy)
2        0.8                                           ( Drama)
3        0.5                                         ( Fantasy)
4        0.3                                          ( Horror)
..       ...                                                ...
140      0.2  ( Drama,  Sci-fi,  Comedy, Action,  Supernatural)
141      0.2  ( Drama,  Shounen,  Comedy, Action,  Supernatu...
142      0.2  ( Drama, Action,  Shounen,  Fantasy,  Supernat...
143      0.2  ( Drama,  Horror,  Shounen, Action,  Supernatu...
144      0.2  ( Drama,  Shounen,  School Life, Action,  Supe...

[145 rows x 2 columns]

With the SUPPORT values above, you can try and match your own manual calculation. You notice how some are grouped with two. ALL items have been tested alone and in groups, but only these ones with the SUPPORT threshold 0.2 or higher values (i.e. other groups didn't make 0.2 therefore not shown).

In [ ]:
#Let's view our interpretation values using the Association rule function.
df_ar=association_rules(df,metric="lift", min_threshold=1)
df_ar

antecedents                                      consequents  \
0       ( Adventure)                                       ( Fantasy)   
1         ( Fantasy)                                     ( Adventure)   
2       ( Adventure)                                        ( Seinen)   
3          ( Seinen)                                     ( Adventure)   
4       ( Adventure)                                         (Action)   
..               ...                                              ...   
901         ( Drama)  ( Shounen,  School Life, Action,  Supernatural)   
902       ( Shounen)    ( School Life,  Drama, Action,  Supernatural)   
903   ( School Life)        ( Shounen,  Drama, Action,  Supernatural)   
904         (Action)  ( Shounen,  School Life,  Drama,  Supernatural)   
905  ( Supernatural)         ( Shounen,  School Life,  Drama, Action)   

     antecedent support  consequent support  support  confidence      lift  \
0                   0.6                 0.5      0.4    0.666667  1.333333   
1                   0.5                 0.6      0.4    0.800000  1.333333   
2                   0.6                 0.2      0.2    0.333333  1.666667   
3                   0.2                 0.6      0.2    1.000000  1.666667   
4                   0.6                 1.0      0.6    1.000000  1.000000   
..                  ...                 ...      ...         ...       ...   
901                 0.8                 0.2      0.2    0.250000  1.250000   
902                 0.7                 0.2      0.2    0.285714  1.428571   
903                 0.2                 0.4      0.2    1.000000  2.500000   
904                 1.0                 0.2      0.2    0.200000  1.000000   
905                 0.6                 0.2      0.2    0.333333  1.666667   

     leverage  conviction  
0        0.10    1.500000  
1        0.10    2.000000  
2        0.08    1.200000  
3        0.08         inf  
4        0.00         inf  
..        ...         ...  
901      0.04    1.066667  
902      0.06    1.120000  
903      0.12         inf  
904      0.00    1.000000  
905      0.08    1.200000  

[906 rows x 9 columns]

From above result, we say that the support is 0.4 calculated by dividing the number of tags containing Adventure and Fantasy by the total number of tags.  The confidence level is 0.67 shows that out of all the tags that contain Adventure 0.67 contain Fantasy too.  The lift 1.33 tells us that Fantasy is 1.23 times more likely to be present on Mangas with the Adventure tag.

In [ ]:
#Let's view our interpretation values using the Association rule function.
df_ar=association_rules(df,metric="confidence", min_threshold=0.6)
df_ar


antecedents                                consequents  \
0                     ( Adventure)                                   ( Drama)   
1                     ( Adventure)                                 ( Fantasy)   
2                       ( Fantasy)                               ( Adventure)   
3                        ( Seinen)                               ( Adventure)   
4                     ( Adventure)                                   (Action)   
..                             ...                                        ...   
504         ( School Life,  Drama)          ( Shounen, Action,  Supernatural)   
505       ( Shounen,  School Life)            ( Drama, Action,  Supernatural)   
506         ( School Life, Action)          ( Shounen,  Drama,  Supernatural)   
507  ( School Life,  Supernatural)                 ( Shounen,  Drama, Action)   
508                 ( School Life)  ( Shounen,  Drama, Action,  Supernatural)   

     antecedent support  consequent support  support  confidence      lift  \
0                   0.6                 0.8      0.4    0.666667  0.833333   
1                   0.6                 0.5      0.4    0.666667  1.333333   
2                   0.5                 0.6      0.4    0.800000  1.333333   
3                   0.2                 0.6      0.2    1.000000  1.666667   
4                   0.6                 1.0      0.6    1.000000  1.000000   
..                  ...                 ...      ...         ...       ...   
504                 0.2                 0.4      0.2    1.000000  2.500000   
505                 0.2                 0.5      0.2    1.000000  2.000000   
506                 0.2                 0.4      0.2    1.000000  2.500000   
507                 0.2                 0.6      0.2    1.000000  1.666667   
508                 0.2                 0.4      0.2    1.000000  2.500000   

     leverage  conviction  
0       -0.08         0.6  
1        0.10         1.5  
2        0.10         2.0  
3        0.08         inf  
4        0.00         inf  
..        ...         ...  
504      0.12         inf  
505      0.10         inf  
506      0.12         inf  
507      0.08         inf  
508      0.12         inf  

[509 rows x 9 columns]

From above result, let's take a look at the first row items: Adventure and Drama. 
The two items have the support of 0.4, where Adventure is 0.6 and Drama is 0.8. 
confidence is 0.67, this implies that if a Manga had Adventure then there is 67% drama is also a tag used. There are also positive correlations as we can see from the positive values of lift, and conviction.

Using this information from ML we can see that Adventure and Drama make a majority of genre tags used in Manga. This leads to a compeling story that captures the aduience attention. That leads to successful Story that will have success in the large market that is Mangas. 